cmd/waku/relay.go

```diff
@@ -0,0 +1,147 @@
+			// Register the node in rendezvous point
+			iter := rendezvous.NewRendezvousPointIterator(options.Rendezvous.Nodes)
+
+			wg.Add(1)
+			go func(nodeTopic string) {
+				t := time.NewTicker(rendezvous.RegisterDefaultTTL)
+				defer t.Stop()
+				defer wg.Done()
+
+				for {
+					select {
+					case <-ctx.Done():
+						return
+					case <-t.C:
+						// Register in rendezvous points periodically
+						wakuNode.Rendezvous().RegisterWithNamespace(ctx, nodeTopic, iter.RendezvousPoints())
+					}
+				}
+			}(nodeTopic)
+
+			wg.Add(1)
+			go func(nodeTopic string) {
+				defer wg.Done()
+				desiredOutDegree := wakuNode.Relay().Params().D
+				t := time.NewTicker(7 * time.Second)
+				defer t.Stop()
+				for {
+					select {
+					case <-ctx.Done():
+						return
+					case <-t.C:
+						peerCnt := len(wakuNode.Relay().PubSub().ListPeers(nodeTopic))
+						peersToFind := desiredOutDegree - peerCnt
+						if peersToFind <= 0 {
+							continue
+						}
+
+						rp := <-iter.Next(ctx)
+						if rp == nil {
+							continue
+						}
+						ctx, cancel := context.WithTimeout(ctx, 7*time.Second)
+						wakuNode.Rendezvous().DiscoverWithNamespace(ctx, nodeTopic, rp, peersToFind)
+						cancel()
+					}
+				}
+			}(nodeTopic)
+
+		}
+	}
+
+	// Bridge topics
+	bridgedTopics := make(map[string]map[string]struct{})
+	bridgedTopicsSet := make(map[string]struct{})
+	for _, topics := range options.Relay.BridgeTopics {
+		_, ok := bridgedTopics[topics.FromTopic]
+		if !ok {
+			bridgedTopics[topics.FromTopic] = make(map[string]struct{})
+		}
+
+		bridgedTopics[topics.FromTopic][topics.ToTopic] = struct{}{}
+		bridgedTopicsSet[topics.FromTopic] = struct{}{}
+		bridgedTopicsSet[topics.ToTopic] = struct{}{}
+	}
+
+	// Make sure all topics are subscribed
+	for _, topic := range maps.Keys(bridgedTopicsSet) {
+		if !wakuNode.Relay().IsSubscribed(topic) {
+			_, err := wakuNode.Relay().Subscribe(ctx, wprotocol.NewContentFilter(topic), relay.WithoutConsumer())
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	for fromTopic, toTopics := range bridgedTopics {
+		subscriptions, err := wakuNode.Relay().Subscribe(ctx, wprotocol.NewContentFilter(fromTopic))
+		if err != nil {
+			return err
+		}
+
+		topics := maps.Keys(toTopics)
+		for _, subscription := range subscriptions {
+			wg.Add(1)
+			go func(subscription *relay.Subscription, topics []string) {
+				defer wg.Done()
+				for env := range subscription.Ch {
+					for _, topic := range topics {
+						// HACK: message has been already fwded
+						metaLen := len(env.Message().Meta)
+						fwdTagLen := len(fwdMetaTag)
+						if metaLen >= fwdTagLen && bytes.Equal(env.Message().Meta[metaLen-fwdTagLen:], fwdMetaTag) {
+							continue
+						}
+
+						// HACK: We append magic numbers here, just so the pubsub message ID will change
+						env.Message().Meta = append(env.Message().Meta, fwdMetaTag...)
+						_, err := wakuNode.Relay().Publish(ctx, env.Message(), relay.WithPubSubTopic(topic))
```

Comment:
If we run this bridging logic in all fleet nodes, I am wondering if duplicates of same messages would get forwarded by each node onto the other topics. 
Not sure about this, but does messageID generation include the originating peerID which has published message into the pubsub? If so, then this would be a problem which needs to be solved.

Reply:
Messages sent via relay will only have the payload and the topic; gossipsub was configured to not include signaturesm peer information or sequence numbers. So it should be okay, nodes that have already seen a messageID for some message they receive would not forward this message further